script/script.go

```diff
@@ -62,72 +130,92 @@ func (ls *luaScript) CreateFilter(config []interface{}) (filters.Filter, error)
 		params = append(params, ps)
 	}
 
-	s := &script{source: src, routeParams: params}
+	s := &script{
+		source:      src,
+		globalPool:  ls.pool,
+		routeParams: params,
+	}
 	if err := s.initScript(); err != nil {
 		return nil, err
 	}
 	return s, nil
 }
 
-func (s *script) getState() (*lua.LState, error) {
-	select {
-	case L := <-s.pool:
-		if L == nil {
-			return nil, errors.New("pool closed")
-		}
-		return L, nil
-	default:
-		return s.newState()
-	}
-}
+func createScript(L *lua.LState, proto *lua.FunctionProto) (*lua.LState, error) {
+	L.Push(L.NewFunctionFromProto(proto))
 
-func (s *script) putState(L *lua.LState) {
-	if s.pool == nil { // pool closed
-		L.Close()
-		return
-	}
-	select {
-	case s.pool <- L:
-	default: // pool full, close state
+	err := L.PCall(0, lua.MultRet, nil)
+	if err != nil {
 		L.Close()
+		return nil, err
 	}
+	return L, nil
+}
+
+type script struct {
+	source                  string
+	routeParams             []string
+	globalPool              *statePool
+	proto                   *lua.FunctionProto
+	hasRequest, hasResponse bool
 }
 
 func (s *script) newState() (*lua.LState, error) {
-	L := lua.NewState()
-	L.PreloadModule("base64", base64.Loader)
-	L.PreloadModule("http", gluahttp.NewHttpModule(&http.Client{}).Loader)
-	L.PreloadModule("url", gluaurl.Loader)
-	L.PreloadModule("json", gjson.Loader)
-	L.SetGlobal("print", L.NewFunction(printToLog))
-	L.SetGlobal("sleep", L.NewFunction(sleep))
+	L, err := s.globalPool.GetState()
+	if err != nil {
+		return nil, err
+	}
 
+	// put compiled script into the current LState
 	L.Push(L.NewFunctionFromProto(s.proto))
 
-	err := L.PCall(0, lua.MultRet, nil)
+	err = L.PCall(0, lua.MultRet, nil)
```

Comment:
We need to confirm somehow that it is ok to re-run possibly different pre-compiled scripts over the same lua state. The [docs](https://github.com/yuin/gopher-lua#sharing-lua-byte-code-between-lstates) only mention sharing the same code with many states (1:N) whereas this approach is M:N

Reply:
That's something we need to do before working too much more on this.
In my tests the results looked great.